File: frontend/src/core/project-workflows/json-schema.ts
import { serializeCanonicalJsonValue } from "./canonical";
import { sha256Utf8 } from "./sha256";
import type { JsonSchema, JsonValue, WorkflowValueType } from "./types";

export const INLINE_SCHEMA_REF_PREFIX =
  "inline-json-schema-v1:sha256:" as const;
export const MAX_JSON_SCHEMA_DEPTH = 16;
export const MAX_JSON_SCHEMA_PROPERTIES = 256;
export const MAX_JSON_SCHEMA_ITEMS = 10_000;
export const MAX_JSON_SCHEMA_ENUM_VALUES = 256;

const primitiveTypes = new Set([
  "null",
  "boolean",
  "object",
  "array",
  "number",
  "integer",
  "string",
]);
const allowedKeywords = new Set([
  "type",
  "properties",
  "required",
  "additionalProperties",
  "items",
  "minItems",
  "maxItems",
  "minLength",
  "maxLength",
  "minimum",
  "maximum",
  "exclusiveMinimum",
  "exclusiveMaximum",
  "enum",
  "const",
  "default",
  "title",
  "description",
  "anyOf",
]);

const objectValue = (value: JsonValue | undefined): JsonSchema | undefined =>
  value !== undefined &&
  value !== null &&
  typeof value === "object" &&
  !Array.isArray(value)
    ? value
    : undefined;

const typeNames = (schema: JsonSchema): string[] => {
  const declared = schema.type;
  let names: string[];
  if (typeof declared === "string") names = [declared];
  else if (
    Array.isArray(declared) &&
    declared.length > 0 &&
    declared.every((item) => typeof item === "string")
  )
    names = declared;
  else if (Array.isArray(schema.anyOf) && schema.anyOf.length > 0) {
    names = schema.anyOf.flatMap((alternative) => {
      const child = objectValue(alternative);
      if (child === undefined)
        throw new Error("anyOf alternatives must be object schemas");
      return typeNames(child);
    });
  } else throw new Error("a closed output schema must declare one type");
  if (names.some((name) => !primitiveTypes.has(name)))
    throw new Error("schema type is outside the frozen primitive set");
  if (new Set(names).size !== names.length)
    throw new Error("schema type alternatives must be unique");
  return names;
};

const typeShape = (schema: JsonSchema) => {
  const names = typeNames(schema);
  const nonNull = names
    .filter((name) => name !== "null")
    .map((name) => (name === "integer" ? "number" : name));
  if (new Set(nonNull).size !== 1 || nonNull.length !== 1)
    throw new Error("schema must contain exactly one non-null top-level type");
  return { nonNullType: nonNull[0]!, nullable: names.includes("null") };
};

const validateSchemaNode = (schema: JsonSchema, depth: number): void => {
  if (depth > MAX_JSON_SCHEMA_DEPTH)
    throw new Error("JSON Schema nesting exceeds the compiler limit");
  const unknown = Object.keys(schema).filter(
    (keyword) => !allowedKeywords.has(keyword),
  );
  if (unknown.length > 0)
    throw new Error(`unsupported JSON Schema keyword: ${unknown.sort()[0]}`);
  const names = typeNames(schema);
  const nonNullTypes = names
    .filter((name) => name !== "null")
    .map((name) => (name === "integer" ? "number" : name));
  if (new Set(nonNullTypes).size > 1)
    throw new Error("multiple non-null schema alternatives are unsupported");
  const nonNullType = nonNullTypes[0];
  if (Array.isArray(schema.anyOf)) {
    const siblings = Object.keys(schema).filter(
      (key) => !["anyOf", "title", "description", "default"].includes(key),
    );
    if (siblings.length > 0)
      throw new Error(
        "anyOf cannot be combined with sibling validation keywords",
      );
    for (const alternative of schema.anyOf) {
      const child = objectValue(alternative);
      if (child === undefined)
        throw new Error("anyOf alternatives must be object schemas");
      validateSchemaNode(child, depth + 1);
    }
    return;
  }

  const properties = objectValue(schema.properties);
  if (schema.properties !== undefined && properties === undefined)
    throw new Error("properties must be an object");
  if (properties !== undefined) {
    if (nonNullType !== "object")
      throw new Error("properties is allowed only on object schemas");
    if (Object.keys(properties).length > MAX_JSON_SCHEMA_PROPERTIES)
      throw new Error("JSON Schema properties exceed the compiler limit");
    for (const child of Object.values(properties)) {
      const childSchema = objectValue(child);
      if (childSchema === undefined)
        throw new Error("property schemas must be objects");
      validateSchemaNode(childSchema, depth + 1);
    }
  }
  if (schema.required !== undefined) {
    if (
      nonNullType !== "object" ||
      !Array.isArray(schema.required) ||
      !schema.required.every((item) => typeof item === "string")
    )
      throw new Error("required must be a string array on an object schema");
    const required = schema.required;
    if (
      new Set(required).size !== required.length ||
      required.some((key) => properties?.[key] === undefined)
    )
      throw new Error("required entries must be unique declared properties");
  }
  if (
    schema.additionalProperties !== undefined &&
    (nonNullType !== "object" ||
      typeof schema.additionalProperties !== "boolean")
  )
    throw new Error("additionalProperties must be boolean on an object schema");
  if (schema.items !== undefined) {
    const items = objectValue(schema.items);
    if (nonNullType !== "array" || items === undefined)
      throw new Error("items must be one object schema on an array");
    validateSchemaNode(items, depth + 1);
  }
  for (const [minimumKey, maximumKey, expectedType] of [
    ["minItems", "maxItems", "array"],
    ["minLength", "maxLength", "string"],
  ] as const) {
    const minimum = schema[minimumKey];
    const maximum = schema[maximumKey];
    for (const value of [minimum, maximum]) {
      if (
        value !== undefined &&
        (typeof value !== "number" ||
          !Number.isSafeInteger(value) ||
          value < 0 ||
          value > MAX_JSON_SCHEMA_ITEMS)
      )
        throw new Error(`${minimumKey}/${maximumKey} exceeds the limit`);
    }
    if (
      (minimum !== undefined || maximum !== undefined) &&
      nonNullType !== expectedType
    )
      throw new Error(`${minimumKey}/${maximumKey} has the wrong type`);
    if (
      typeof minimum === "number" &&
      typeof maximum === "number" &&
      minimum > maximum
    )
      throw new Error(`${minimumKey}/${maximumKey} bounds must be ordered`);
  }
  for (const key of [
    "minimum",
    "maximum",
    "exclusiveMinimum",
    "exclusiveMaximum",
  ] as const) {
    const value = schema[key];
    if (value !== undefined) {
      if (nonNullType !== "number")
        throw new Error(`${key} is allowed only on number schemas`);
      if (typeof value !== "number" || !Number.isFinite(value))
        throw new Error(`${key} must be a finite number`);
    }
  }
  if (
    typeof schema.minimum === "number" &&
    typeof schema.maximum === "number" &&
    schema.minimum > schema.maximum
  )
    throw new Error("numeric bounds must be ordered");
  if (
    schema.enum !== undefined &&
    (!Array.isArray(schema.enum) ||
      schema.enum.length === 0 ||
      schema.enum.length > MAX_JSON_SCHEMA_ENUM_VALUES)
  )
    throw new Error("enum must contain a bounded non-empty value set");
  if (Array.isArray(schema.enum)) {
    const canonical = schema.enum.map(serializeCanonicalJsonValue);
    if (new Set(canonical).size !== canonical.length)
      throw new Error("enum values must be unique");
  }
  for (const key of ["title", "description"] as const) {
    if (schema[key] !== undefined && typeof schema[key] !== "string")
      throw new Error(`${key} must be a string`);
  }
};

const valueMatchesSchema = (value: JsonValue, schema: JsonSchema): boolean => {
  if (Array.isArray(schema.anyOf)) {
    return schema.anyOf.some((alternative) => {
      const child = objectValue(alternative);
      return child !== undefined && valueMatchesSchema(value, child);
    });
  }
  const names = typeNames(schema);
  const matchesDeclaredType = names.some((name) => {
    if (name === "null") return value === null;
    if (name === "boolean") return typeof value === "boolean";
    if (name === "string") return typeof value === "string";
    if (name === "number")
      return typeof value === "number" && Number.isFinite(value);
    if (name === "integer")
      return typeof value === "number" && Number.isSafeInteger(value);
    if (name === "array") return Array.isArray(value);
    return value !== null && typeof value === "object" && !Array.isArray(value);
  });
  if (!matchesDeclaredType) return false;
  const encoded = serializeCanonicalJsonValue(value);
  if (
    Array.isArray(schema.enum) &&
    !schema.enum.some(
      (candidate) => serializeCanonicalJsonValue(candidate) === encoded,
    )
  )
    return false;
  if (
    schema.const !== undefined &&
    serializeCanonicalJsonValue(schema.const) !== encoded
  )
    return false;
  if (typeof value === "string") {
    if (typeof schema.minLength === "number" && value.length < schema.minLength)
      return false;
    if (typeof schema.maxLength === "number" && value.length > schema.maxLength)
      return false;
  }
  if (typeof value === "number") {
    if (typeof schema.minimum === "number" && value < schema.minimum)
      return false;
    if (typeof schema.maximum === "number" && value > schema.maximum)
      return false;
    if (
      typeof schema.exclusiveMinimum === "number" &&
      value <= schema.exclusiveMinimum
    )
      return false;
    if (
      typeof schema.exclusiveMaximum === "number" &&
      value >= schema.exclusiveMaximum
    )
      return false;
  }
  if (Array.isArray(value)) {
    if (typeof schema.minItems === "number" && value.length < schema.minItems)
      return false;
    if (typeof schema.maxItems === "number" && value.length > schema.maxItems)
      return false;
    const items = objectValue(schema.items);
    if (
      items !== undefined &&
      !value.every((item) => valueMatchesSchema(item, items))
    )
      return false;
  } else if (value !== null && typeof value === "object") {
    const properties = objectValue(schema.properties) ?? {};
    const required = Array.isArray(schema.required)
      ? schema.required.filter((key): key is string => typeof key === "string")
      : [];
    if (required.some((key) => !(key in value))) return false;
    for (const [key, childValue] of Object.entries(value)) {
      const childSchema = objectValue(properties[key]);
      if (childSchema !== undefined) {
        if (!valueMatchesSchema(childValue, childSchema)) return false;
      } else if (schema.additionalProperties === false) return false;
    }
  }
  return true;
};

export const validateStrictJsonSchema = (schema: JsonSchema): void => {
  validateSchemaNode(schema, 1);
  typeShape(schema);
  if (
    schema.default !== undefined &&
    !valueMatchesSchema(schema.default, schema)
  )
    throw new Error("schema default does not satisfy the schema");
  if (schema.const !== undefined && !valueMatchesSchema(schema.const, schema))
    throw new Error("schema const does not satisfy the schema");
};

export const inlineJsonSchemaRef = (schema: JsonSchema): string => {
  validateStrictJsonSchema(schema);
  return `${INLINE_SCHEMA_REF_PREFIX}${sha256Utf8(
    serializeCanonicalJsonValue(schema),
  )}`;
};

export const valueTypeFromJsonSchema = (
  schema: JsonSchema,
  requirement: "any" | "object" = "any",
): WorkflowValueType => {
  validateStrictJsonSchema(schema);
  const shape = typeShape(schema);
  if (
    requirement === "object" &&
    (shape.nonNullType !== "object" || shape.nullable)
  )
    throw new Error("this node output requires a non-null object schema");
  const kind =
    shape.nonNullType === "string"
      ? "string"
      : shape.nonNullType === "number"
        ? "number"
        : shape.nonNullType === "boolean"
          ? "boolean"
          : "json";
  return {
    kind,
    collection: shape.nonNullType === "array",
    nullable: shape.nullable,
    schema_ref: inlineJsonSchemaRef(schema),
  };
};

File: backend/packages/harness/deerflow/workflows/json_schema.py
"""Closed JSON Schema subset used by first-batch typed node outputs.

The subset is deliberately smaller than Draft 2020-12.  It has no references,
regular expressions, conditionals, or open-ended applicators, so Registry port
derivation and JSON Pointer typing stay deterministic across Python/TypeScript.
"""

from __future__ import annotations

import hashlib
from dataclasses import dataclass
from typing import Final, Literal

from jsonschema import Draft202012Validator, SchemaError

from deerflow.workflows.canonical import canonical_json_value
from deerflow.workflows.contracts import JsonSchema, WorkflowValueType

INLINE_SCHEMA_REF_PREFIX: Final = "inline-json-schema-v1:sha256:"
MAX_JSON_SCHEMA_DEPTH: Final = 16
MAX_JSON_SCHEMA_PROPERTIES: Final = 256
MAX_JSON_SCHEMA_ITEMS: Final = 10_000
MAX_JSON_SCHEMA_ENUM_VALUES: Final = 256

_PRIMITIVE_TYPES: Final = frozenset({"null", "boolean", "object", "array", "number", "integer", "string"})
_ALLOWED_KEYWORDS: Final = frozenset(
    {
        "type",
        "properties",
        "required",
        "additionalProperties",
        "items",
        "minItems",
        "maxItems",
        "minLength",
        "maxLength",
        "minimum",
        "maximum",
        "exclusiveMinimum",
        "exclusiveMaximum",
        "enum",
        "const",
        "default",
        "title",
        "description",
        "anyOf",
    }
)


class WorkflowJsonSchemaError(ValueError):
    """An authored schema is outside the closed compiler-v1 subset."""

    def __init__(self, code: str, message: str) -> None:
        self.code = code
        super().__init__(message)


@dataclass(frozen=True, slots=True)
class JsonSchemaTypeShape:
    non_null_type: Literal["boolean", "object", "array", "number", "string"]
    nullable: bool


def _invalid(message: str) -> WorkflowJsonSchemaError:
    return WorkflowJsonSchemaError("WORKFLOW_JSON_SCHEMA_INVALID", message)


def _unsupported(message: str) -> WorkflowJsonSchemaError:
    return WorkflowJsonSchemaError("WORKFLOW_JSON_SCHEMA_UNSUPPORTED_KEYWORD", message)


def _limited(message: str) -> WorkflowJsonSchemaError:
    return WorkflowJsonSchemaError("WORKFLOW_JSON_SCHEMA_LIMIT_EXCEEDED", message)


def _type_names(schema: JsonSchema) -> tuple[str, ...]:
    declared = schema.get("type")
    if isinstance(declared, str):
        names = (declared,)
    elif isinstance(declared, list) and declared and all(isinstance(item, str) for item in declared):
        names = tuple(declared)
    elif "anyOf" in schema:
        alternatives = schema["anyOf"]
        if not isinstance(alternatives, list) or not alternatives:
            raise _invalid("anyOf must contain at least one schema")
        names_list: list[str] = []
        for alternative in alternatives:
            if not isinstance(alternative, dict):
                raise _invalid("anyOf alternatives must be object schemas")
            names_list.extend(_type_names(alternative))
        names = tuple(names_list)
    else:
        raise _invalid("a closed output schema must declare one explicit type")
    if any(name not in _PRIMITIVE_TYPES for name in names):
        raise _invalid("schema type is outside the frozen primitive set")
    if len(names) != len(set(names)):
        raise _invalid("schema type alternatives must be unique")
    return names


def json_schema_type_shape(schema: JsonSchema) -> JsonSchemaTypeShape:
    names = _type_names(schema)
    nullable = "null" in names
    non_null = [name for name in names if name != "null"]
    if len(non_null) != 1:
        raise WorkflowJsonSchemaError(
            "WORKFLOW_JSON_SCHEMA_TOP_LEVEL_TYPE_INVALID",
            "schema must contain exactly one non-null top-level type",
        )
    normalized = "number" if non_null[0] == "integer" else non_null[0]
    return JsonSchemaTypeShape(non_null_type=normalized, nullable=nullable)  # type: ignore[arg-type]


def _validate_schema_node(schema: JsonSchema, *, depth: int) -> None:
    if depth > MAX_JSON_SCHEMA_DEPTH:
        raise _limited("JSON Schema nesting exceeds the compiler limit")
    unknown = set(schema) - _ALLOWED_KEYWORDS
    if unknown:
        raise _unsupported(f"unsupported JSON Schema keyword: {min(unknown)}")
    if "anyOf" in schema and any(
        key in schema
        for key in (
            "type",
            "properties",
            "required",
            "additionalProperties",
            "items",
            "minItems",
            "maxItems",
            "minLength",
            "maxLength",
            "minimum",
            "maximum",
            "exclusiveMinimum",
            "exclusiveMaximum",
            "enum",
            "const",
        )
    ):
        raise _invalid("anyOf cannot be combined with sibling validation keywords")

    names = _type_names(schema)
    non_null_names = {"number" if name == "integer" else name for name in names if name != "null"}
    if len(non_null_names) > 1:
        raise WorkflowJsonSchemaError(
            "WORKFLOW_JSON_SCHEMA_TOP_LEVEL_TYPE_INVALID",
            "multiple non-null schema alternatives are unsupported",
        )

    alternatives = schema.get("anyOf")
    if isinstance(alternatives, list):
        for alternative in alternatives:
            assert isinstance(alternative, dict)
            _validate_schema_node(alternative, depth=depth + 1)
        return

    properties = schema.get("properties")
    if properties is not None:
        if non_null_names != {"object"} or not isinstance(properties, dict):
            raise _invalid("properties is allowed only on object schemas")
        if len(properties) > MAX_JSON_SCHEMA_PROPERTIES:
            raise _limited("JSON Schema properties exceed the compiler limit")
        for child in properties.values():
            if not isinstance(child, dict):
                raise _invalid("property schemas must be objects")
            _validate_schema_node(child, depth=depth + 1)
    required = schema.get("required")
    if required is not None:
        if non_null_names != {"object"} or not isinstance(required, list) or not all(isinstance(item, str) for item in required):
            raise _invalid("required must be a string array on an object schema")
        if len(required) != len(set(required)) or any(item not in (properties or {}) for item in required):
            raise _invalid("required entries must be unique declared properties")
    additional = schema.get("additionalProperties")
    if additional is not None and (non_null_names != {"object"} or type(additional) is not bool):
        raise _invalid("additionalProperties must be boolean on an object schema")

    items = schema.get("items")
    if items is not None:
        if non_null_names != {"array"} or not isinstance(items, dict):
            raise _invalid("items must be one object schema on an array")
        _validate_schema_node(items, depth=depth + 1)
    for minimum_key, maximum_key, schema_type, upper_bound in (
        ("minItems", "maxItems", "array", MAX_JSON_SCHEMA_ITEMS),
        ("minLength", "maxLength", "string", MAX_JSON_SCHEMA_ITEMS),
    ):
        minimum = schema.get(minimum_key)
        maximum = schema.get(maximum_key)
        for value in (minimum, maximum):
            if value is not None and (type(value) is not int or value < 0 or value > upper_bound):
                raise _limited(f"{minimum_key}/{maximum_key} exceeds the compiler limit")
        if (minimum is not None or maximum is not None) and non_null_names != {schema_type}:
            raise _invalid(f"{minimum_key}/{maximum_key} is invalid for this schema type")
        if minimum is not None and maximum is not None and minimum > maximum:
            raise _invalid(f"{minimum_key}/{maximum_key} bounds must be ordered")

    for key in ("minimum", "maximum", "exclusiveMinimum", "exclusiveMaximum"):
        if key in schema and non_null_names != {"number"}:
            raise _invalid(f"{key} is allowed only on number schemas")
    if "minimum" in schema and "maximum" in schema and schema["minimum"] > schema["maximum"]:  # type: ignore[operator]
        raise _invalid("numeric bounds must be ordered")

    enum = schema.get("enum")
    if enum is not None:
        if not isinstance(enum, list) or not enum or len(enum) > MAX_JSON_SCHEMA_ENUM_VALUES:
            raise _limited("enum must contain a bounded non-empty value set")
        canonical = [canonical_json_value(item) for item in enum]
        if len(canonical) != len(set(canonical)):
            raise _invalid("enum values must be unique")


def validate_strict_json_schema(schema: JsonSchema) -> None:
    try:
        Draft202012Validator.check_schema(schema)
    except SchemaError as error:
        raise _invalid("JSON Schema is not valid Draft 2020-12") from error
    _validate_schema_node(schema, depth=1)
    shape = json_schema_type_shape(schema)
    if "default" in schema and not Draft202012Validator(schema).is_valid(schema["default"]):
        raise _invalid("schema default does not satisfy the schema")
    if "const" in schema and not Draft202012Validator(schema).is_valid(schema["const"]):
        raise _invalid("schema const does not satisfy the schema")
    if shape.non_null_type == "array" and "items" not in schema:
        # Generic arrays are allowed as outputs, but pointer access remains fail-closed.
        return


def inline_json_schema_ref(schema: JsonSchema) -> str:
    validate_strict_json_schema(schema)
    canonical = canonical_json_value(schema).encode("utf-8")
    return INLINE_SCHEMA_REF_PREFIX + hashlib.sha256(canonical).hexdigest()


def value_type_from_json_schema(
    schema: JsonSchema,
    *,
    require_top_level: Literal["any", "object"] = "any",
) -> WorkflowValueType:
    validate_strict_json_schema(schema)
    shape = json_schema_type_shape(schema)
    if require_top_level == "object" and (shape.non_null_type != "object" or shape.nullable):
        raise WorkflowJsonSchemaError(
            "WORKFLOW_JSON_SCHEMA_TOP_LEVEL_TYPE_INVALID",
            "this node output requires a non-null top-level object schema",
        )
    kind = {
        "string": "string",
        "number": "number",
        "boolean": "boolean",
        "object": "json",
        "array": "json",
    }[shape.non_null_type]
    return WorkflowValueType(
        kind=kind,
        collection=shape.non_null_type == "array",
        nullable=shape.nullable,
        schema_ref=inline_json_schema_ref(schema),
    )


def _single_schema(schema: JsonSchema) -> tuple[JsonSchema, bool]:
    shape = json_schema_type_shape(schema)
    alternatives = schema.get("anyOf")
    if not isinstance(alternatives, list):
        return schema, shape.nullable
    for alternative in alternatives:
        assert isinstance(alternative, dict)
        names = _type_names(alternative)
        if any(name != "null" for name in names):
            return alternative, shape.nullable
    raise _invalid("nullable schema is missing its non-null alternative")


def _decode_pointer_token(token: str) -> str:
    result: list[str] = []
    index = 0
    while index < len(token):
        if token[index] != "~":
            result.append(token[index])
            index += 1
            continue
        if index + 1 >= len(token) or token[index + 1] not in {"0", "1"}:
            raise WorkflowJsonSchemaError(
                "WORKFLOW_NODE_OUTPUT_PATH_INVALID",
                "JSON Pointer contains an invalid escape",
            )
        result.append("~" if token[index + 1] == "0" else "/")
        index += 2
    return "".join(result)


def value_type_at_json_pointer(schema: JsonSchema, pointer: str) -> WorkflowValueType:
    validate_strict_json_schema(schema)
    if pointer == "":
        return value_type_from_json_schema(schema)
    if not pointer.startswith("/"):
        raise WorkflowJsonSchemaError(
            "WORKFLOW_NODE_OUTPUT_PATH_INVALID",
            "JSON Pointer must be empty or start with a slash",
        )
    nullable_from_path = False
    current = schema
    for raw_token in pointer[1:].split("/"):
        token = _decode_pointer_token(raw_token)
        current, parent_nullable = _single_schema(current)
        nullable_from_path = nullable_from_path or parent_nullable
        shape = json_schema_type_shape(current)
        if shape.non_null_type == "object":
            properties = current.get("properties")
            if not isinstance(properties, dict) or token not in properties:
                raise WorkflowJsonSchemaError(
                    "WORKFLOW_NODE_OUTPUT_PATH_INVALID",
                    "JSON Pointer does not resolve to a declared property",
                )
            required = current.get("required", [])
            nullable_from_path = nullable_from_path or token not in required
            child = properties[token]
            assert isinstance(child, dict)
            current = child
        elif shape.non_null_type == "array":
            if not token.isdigit() or (len(token) > 1 and token.startswith("0")):
                raise WorkflowJsonSchemaError(
                    "WORKFLOW_NODE_OUTPUT_PATH_INVALID",
                    "array JSON Pointer tokens must be canonical non-negative indexes",
                )
            items = current.get("items")
            if not isinstance(items, dict):
                raise WorkflowJsonSchemaError(
                    "WORKFLOW_NODE_OUTPUT_PATH_INVALID",
                    "array JSON Pointer typing requires one items schema",
                )
            nullable_from_path = True
            current = items
        else:
            raise WorkflowJsonSchemaError(
                "WORKFLOW_NODE_OUTPUT_PATH_INVALID",
                "JSON Pointer cannot traverse a scalar schema",
            )
    result = value_type_from_json_schema(current)
    if nullable_from_path and not result.nullable:
        result = result.model_copy(update={"nullable": True})
    return result


__all__ = [
    "INLINE_SCHEMA_REF_PREFIX",
    "MAX_JSON_SCHEMA_DEPTH",
    "MAX_JSON_SCHEMA_ENUM_VALUES",
    "MAX_JSON_SCHEMA_ITEMS",
    "MAX_JSON_SCHEMA_PROPERTIES",
    "WorkflowJsonSchemaError",
    "inline_json_schema_ref",
    "json_schema_type_shape",
    "validate_strict_json_schema",
    "value_type_at_json_pointer",
    "value_type_from_json_schema",
]
